src/relay/pass/partial_eval.cc

```diff
@@ -0,0 +1,719 @@
+ * and allow us to not handle capture-avoidance substitution (as atom has no binder).
+ *
+ * 2: The map of References to partially static values is Reified, as described below.
+ * Instead of Reference having mutable field, Reference only has an unique identifier.
+ * There will be a mutable mapping of id to partially static value, called the store.
+ * This allow us to rollback the store:
+ * when a path may or may not be executed (as in a conditional), we copy the store,
+ * recursively Reify with the copy, and reinstate the original when the call returns
+ * so that that the effects of the computation are not preserved.
+ * We do this in if else, pattern matching, and in function,
+ * as, when we see a function, we partially evaluate it with all the argument as dynamic,
+ * to generate efficient dynamic for that function.
+ *
+ * 3: The generated code reuses bindings (although they are not shadowed),
+ * so we have to deduplicate them.
+ *
+ * Also, It will also generate lots of dead code,
+ * so it is a good idea to feed it through the dead code eliminator after partial evaluation.
+ *
+ * The partial evaluator makes several assumptions, so there is room for improvement:
+ *
+ * 0: The partial evaluator treats global variables as opaque.
+ * Doing PartialEval on a module level will solve this.
+ *
+ * 1: The partial evaluator assume all functions as terminating.
+ * We need to has a max_expand parameter that shrink on every compile time evaluation,
+ * to make sure PE does not infinite loop.
+ * Additionally, we might add a termination analysis pass that lift this requirement
+ * for function that analysis found terminating.
+ *
+ * 2: Every time an unknown effect happened, we clear the whole store.
+ * It is too conservative: if a local reference is created (and do not get passed outside),
+ * An unknown global function call/global reference write can not modify it.
+ * We can pair PE with escape analysis/alias analysis.
+ *
+ * 3: We assume all unknown code has effect. Doing effect analysis can make the store more precise.
+ *
+ * 4: When doing pattern matching, we can simplify the match even for dynamic case.
+ * Right now it is all or nothing: either a complete match, or the original dynamic code.
+ * Instead, we can get a match tree, pair it with the data and evaluate it to a normal form.
+ * We then can reify the result.
+ *
+ * 5: Every time a function is called, it's code will get expanded and partially evaluated.
+ * We can do a binding time analysis to cache the result and avoid re-partial evaluation.
+ *
+ * These assumptions do not affect the correctness of the algorithm, however.
+ */
+#include <tvm/relay/pass.h>
+#include <tvm/relay/expr_functor.h>
+#include <tvm/relay/pattern_functor.h>
+#include <tvm/relay/interpreter.h>
+#include "pass_util.h"
+#include "let_list.h"
+
+namespace tvm {
+namespace relay {
+
+using namespace runtime;
+
+struct StaticNode {
```

Comment:
Why not just inherit the TVM's node system? Should be similar i guess

Reply:
It is considerably more work, and we has no need.

Reply:
While it might take a bit more work to adopt TVM's node system, it makes us share the same infra and reuse later. I will recommend using node instead here.

Note that it is not too much more work, just replace shared_ptr->NodePtr and make_shared->make_node